rectools/models/nn/transformer_base.py

```diff
@@ -0,0 +1,506 @@
+    tp.Type[TransformerDataPreparatorBase],
+    BeforeValidator(_get_class_obj),
+    PlainSerializer(
+        func=get_class_or_function_full_path,
+        return_type=str,
+        when_used="json",
+    ),
+]
+
+
+ItemNetConstructorType = tpe.Annotated[
+    tp.Type[ItemNetConstructorBase],
+    BeforeValidator(_get_class_obj),
+    PlainSerializer(
+        func=get_class_or_function_full_path,
+        return_type=str,
+        when_used="json",
+    ),
+]
+
+ItemNetBlockTypes = tpe.Annotated[
+    tp.Sequence[tp.Type[ItemNetBase]],
+    BeforeValidator(_get_class_obj_sequence),
+    PlainSerializer(
+        func=_serialize_type_sequence,
+        return_type=str,
+        when_used="json",
+    ),
+]
+
+
+ValMaskCallable = Callable[[], np.ndarray]
+
+ValMaskCallableSerialized = tpe.Annotated[
+    ValMaskCallable,
+    BeforeValidator(_get_class_obj),
+    PlainSerializer(
+        func=get_class_or_function_full_path,
+        return_type=str,
+        when_used="json",
+    ),
+]
+
+TrainerCallable = Callable[[], Trainer]
+
+TrainerCallableSerialized = tpe.Annotated[
+    TrainerCallable,
+    BeforeValidator(_get_class_obj),
+    PlainSerializer(
+        func=get_class_or_function_full_path,
+        return_type=str,
+        when_used="json",
+    ),
+]
+
+
+class TransformerModelConfig(ModelConfig):
+    """Transformer model base config."""
+
+    data_preparator_type: TransformerDataPreparatorType
+    n_blocks: int = 2
+    n_heads: int = 4
+    n_factors: int = 256
+    use_pos_emb: bool = True
+    use_causal_attn: bool = False
+    use_key_padding_mask: bool = False
+    dropout_rate: float = 0.2
+    session_max_len: int = 100
+    dataloader_num_workers: int = 0
+    batch_size: int = 128
+    loss: str = "softmax"
+    n_negatives: int = 1
+    gbce_t: float = 0.2
+    lr: float = 0.001
+    epochs: int = 3
+    verbose: int = 0
+    deterministic: bool = False
+    recommend_batch_size: int = 256
+    recommend_device: tp.Optional[str] = None
+    recommend_n_threads: int = 0
+    recommend_use_torch_ranking: bool = True
+    train_min_user_interactions: int = 2
+    item_net_block_types: ItemNetBlockTypes = (IdEmbeddingsItemNet, CatFeaturesItemNet)
+    item_net_constructor_type: ItemNetConstructorType = SumOfEmbeddingsConstructor
+    pos_encoding_type: PositionalEncodingType = LearnableInversePositionalEncoding
+    transformer_layers_type: TransformerLayersType = PreLNTransformerLayers
+    lightning_module_type: TransformerLightningModuleType = TransformerLightningModule
+    get_val_mask_func: tp.Optional[ValMaskCallableSerialized] = None
+    get_trainer_func: tp.Optional[TrainerCallableSerialized] = None
+
+
+TransformerModelConfig_T = tp.TypeVar("TransformerModelConfig_T", bound=TransformerModelConfig)
+
+
+# ####  --------------  Transformer Model Base  --------------  #### #
+
+
+class TransformerModelBase(ModelBase[TransformerModelConfig_T]):  # pylint: disable=too-many-instance-attributes
```

Comment:
Kind of an orcestrator class for actual data preparation, training and net classes. Doesn't do anything on it's own. But manages other classes and creates a single pipeline out of them. Also all initializations are here.

Configs and checkpoints are also here.

Recommending logic was moved to LightningModule on purpose.
We will not always do just dot product for getting logits. So we need to have this tying in one class. LightningModule has it all now, both for training and recommending.